compiler/resolution/nilChecking.cpp

```diff
@@ -0,0 +1,782 @@
+  std::map<Symbol*, int> symToIdx;
+
+  {
+    int index = 0;
+    std::vector<SymExpr*> symExprs;
+    collectSymExprs(fn, symExprs);
+    for_vector(SymExpr, se, symExprs) {
+      Symbol* sym = se->symbol();
+      if (isArgSymbol(sym) || isVarSymbol(sym)) {
+        if (symToIdx.count(sym) == 0) {
+          // Add it to the map/array
+          symToIdx.insert(std::pair<Symbol*, int>(sym, index));
+          idxToSym.push_back(sym);
+          INT_ASSERT((int)(idxToSym.size()-1) == index);
+          index++;
+          if (debugging) {
+            printf("Tracking symbol %i %s\n", sym->id, sym->name);
+          }
+        }
+      }
+    }
+  }
+
+  size_t nbbs = fn->basicBlocks->size();
+
+  // allocate the IN and OUT data flow sets
+  std::vector<AliasMap> IN(nbbs);
+  std::vector<AliasMap> OUT(nbbs);
+
+  bool changed = false;
+
+  do {
+
+    changed = false;
+
+    for (size_t i = 0; i < nbbs; i++) {
+      if (debugging) {
+        printf("running checkBasicBlock on bb %i\n", (int)i);
+      }
+      checkBasicBlock(fn, (*fn->basicBlocks)[i], idxToSym, IN[i], OUT[i],
+                      /*raise errors?*/ false);
+    }
+
+    // for each basic block, update IN block based on predecessors OUT blocks
+    for (size_t i = 0; i < nbbs; i++) {
+      BasicBlock* bb = (*fn->basicBlocks)[i];
+
+      AliasMap nextIn;
+
+      for_vector(BasicBlock, bbin, bb->ins) {
+        const AliasMap& from = OUT[bbin->id];
+        for (AliasMap::const_iterator it = from.begin();
+             it != from.end();
+             ++it) {
+          Symbol* sym = it->first;
+          AliasLocation loc = it->second;
+          if (nextIn.count(sym) == 0) {
+            nextIn[sym] = loc;
+          } else {
+            AliasLocation was = nextIn[sym];
+            // combine alias locations
+            if (was.type == loc.type &&
+                was.location == loc.location) {
+              // OK, do nothing
+            } else if (was.type == MUST_ALIAS_IGNORED) {
+              nextIn[sym] = loc;
+            } else {
+              // Conflicting input -> unknown
+              nextIn[sym] = unknownAliasLocation();
+            }
+          }
+        }
+      }
+
+      // Now, propagate changes from nextIn to IN[i]
+      // To ensure the algorithm converges, we do this
+      // in a way that makes sure AliasLocations for
+      // a given value are monotonically increasing.
+      AliasMap& to = IN[i];
+      for (AliasMap::const_iterator it = nextIn.begin();
+           it != nextIn.end();
+           ++it) {
+        Symbol* sym = it->first;
+        AliasLocation loc = it->second;
+
+        // might default-initialize it with MUST_ALIAS_IGNORED
+        AliasLocation &dst = to[sym];
+
+        if (dst.type < loc.type) {
+          if (debugging) {
+            printAliasEntry("in was", sym, dst);
+            printAliasEntry("in now", sym, loc);
+          }
+          dst = loc;
+          changed = true;
+        }
+      }
+    }
```

Comment:
Ditto the above.

Or, have a single function for combining and propagating, to make the body of the do-while loop very simple.

Reply:
I'd like to leave this one here because it has to do with the fixed-point algorithm's convergence. I moved the others out so now it should fit in a screen though.